### PR TITLE
Added deprecation for non-Undefined read-under-write mems

### PR DESCRIPTION
**Pull request by azidar**
It turns out that because there is no syntax for declaring non-undefined read-under-write mems in FIRRTL/CHIRRTL, going through that serialization path actually reverts the behavior to Undefined.

This is an opportunity to revisit adding CHIRRTL to the FIRRTL spec and adding this back in, but until then, we should not be generating this from Chisel.

### Contributor Checklist

- [N/A] Did you add Scaladoc to every public function/method?
- [N/A] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [N/A] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

 - removed feature/API

#### API Impact

Users of non-undefined memory behavior will get a warning, and in Chisel 5.0, it will get removed. Likely in the future, we will add back the feature once the backing compiler infra supports it more fully.

#### Backend Code Generation Impact

None.

#### Release Notes

Deprecated generating memories with non-Undefined read-under-write collision semantics, as they would silently get changed to Undefined r-u-w behavior if going through a CHIRRTL serialization step.

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.4.x`, [small] API extension: `3.5.x`, API modification or big change: `3.6.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

File: core/src/main/scala/chisel3/Mem.scala [hunk 1]
```diff
@@ -318,6 +318,9 @@ object SyncReadMem {
     if (compileOptions.declaredTypeMustBeUnbound) {
       requireIsChiselType(t, "memory type")
     }
+    if (ruw != Undefined) {
+      Builder.warning(s"specifying read-under-write behavior other than Undefined is deprecated 3.6 and removed in 5.0")
+    }
```
Comment (seldridge) [inline core/src/main/scala/chisel3/Mem.scala:323]: Leaving this and also deprecating the `apply(Int, T, ReadUnderWrite)` methods would provide a compile-time warning.

**Review by seldridge — APPROVED** [review]
LGTM

Can we also deprecate the apply methods that accept a ReadUnderWrite argument?

File: core/src/main/scala/chisel3/Mem.scala (hunk 1 above)
Comment (azidar) [reply core/src/main/scala/chisel3/Mem.scala:323]: Not sure.. if we are planning on adding it back, I'm not sure its worth going through a round of removing the API then adding it back. Do you feel strongly?

Comment (seldridge) [reply core/src/main/scala/chisel3/Mem.scala:323]: I don't feel strongly.

Comment (azidar) [reply core/src/main/scala/chisel3/Mem.scala:323]: @jackkoenig ?

**Comment by ekiwi** [comment]
It would be great if we could somehow preserve this since some circuits actually are broken if read-under write is undefined. See: https://github.com/freechipsproject/ip-contributions/pull/27

If you want to get rid of the options, then it would be good to at least make the memories WriteFirst or ReadFirst instead of undefined. Whatever matches the behavior of the default Verilog that gets generated by firrtl, since many have come to rely on that and if we simulate true undefined read-under-write (e.g. the way the formal backend does it) many circuits will probably be broken.

**Comment by azidar** [comment]
That's a good point @ekiwi. Making it not undefined but the default behavior seems a sensible non-blocking step forwards.

I think we will need to add this back to Chisel post the chisel5 release because I feel like specifying the r-u-w behavior will be necessary, and I suppose from the perspective of the vast majority of users who used the standard ChiselStage, this feature worked for them previously.

@seldridge do you know what the default emitted r-u-w behavior is?

**Comment by seldridge** [comment]
Testing for FIRRTL memory constructs, not CHIRRTL: SFC emits "new" the same as "undefined". MFC emits everything the same (This is a bug).

It's correct to lock in undefined as the only option and immediately makes SFC, MFC, and the FIRRTL serialization correct.  My primary concern is that this is broken for the serialization format (SFC is broken here, too).  Locking in "new" is problematic as it's not reflected in the serialization and FIRRTL compilers can legally miscompile it.

Mid-term multi-part fix:

1. Define a serialization format for read-under-write for CHIRRTL.
2. Have MFC reject "old".
3. Modify MFC to support "old".
4. Modify Chisel to emit CHIRRTL with read-under-write specified.
5. Add CHIRRTL mems to the spec.
6. Change MFC to emit actual undefined read-under-write.

Tl;dr: I think this will come back.  There's just a number of steps to bring it back all centered around serializing this in the IR.

**Comment by ekiwi** [comment]
I fixed the serialization: https://github.com/chipsalliance/firrtl/pull/2595

Turns out that the `firrtl` parser already supported a serialization format. It was just missing from the Serializer.

**Comment by seldridge** [comment]
Awesome. This should be enough to proceed without deprecating this.

**Comment by seldridge** [comment]
Tracking issue to add this to CIRCT: https://github.com/llvm/circt/issues/4588

**Comment by azidar** [comment]
Thanks @ekiwi and @seldridge , I'll close the ticket.
